### PR TITLE
feat(sdk-lib-mpc): support DKLS DKG primitives

### PR DESCRIPTION
**Pull request by islamaminBitGo**
Ticket: HSM-267

1) Provides a wrapper class + types that abstract memory handling with wasm, message handling, and creation of commitments.
2) Provides functions and types to handle authenticated encryption between parties involved in the protocol.
          * Broadcast messages are only authenticated (signed with the sender's private gpg key).
          * P2P messages are authenticated using the sender's private gpg key and encrypted to the recipient's public gpg key).

### CODE REVIEW

File: modules/sdk-core/src/bitgo/utils/opengpgUtils.ts [hunk 1]
```diff
@@ -309,75 +309,6 @@ export async function encryptText(text: string, key: Key): Promise<string> {
   });
 }
 
-/**
- * Encrypts and detach signs a string
- * @param text string to encrypt and sign
- * @param publicArmor public key to encrypt with
- * @param privateArmor private key to sign with
- */
-export async function encryptAndDetachSignText(
```
Comment (islamaminBitGo) [inline modules/sdk-core/src/bitgo/utils/opengpgUtils.ts:318]: i recently added these openpgp utility functions but decided to move them to where they're actually needed plus change them to encrypt + decrypt and sign + verify binary data to be compatible with hsm.

**Review by mmcshinsky-bitgo — CHANGES_REQUESTED** [review]
removal of a publicly exported function is a breaking change. Commit needs to reflect that with BREAKING CHANGE: <description> in the commit footer

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/types.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+export type PartyGpgKey = {
+  partyId: number;
+  gpgKey: string;
+};
+export type SerializedBroadcastMessage = BroadcastMessage<string>;
+export type DeserializedBroadcastMessage = BroadcastMessage<Uint8Array>;
+export type SerializedP2PMessage = P2PMessage<string, string>;
+export type DeserializedP2PMessage = P2PMessage<Uint8Array, Uint8Array>;
+export type AuthEncP2PMessage = P2PMessage<AuthEncMessage, string>;
+export type AuthBroadcastMessage = BroadcastMessage<AuthEncMessage>;
```
Comment (zhongxishen) [inline modules/sdk-lib-mpc/src/tss/ecdsa-dkls/types.ts:38]: `AuthBroadcastMessage` being parameterized to `AuthEncMessage` is a little confusing, maybe apart from `AuthEncMessage`, we need another type `AuthMessage`.

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    format: 'armored',
+    detached: true,
+    config: {
+      rejectCurves: new Set(),
+      showVersion: false,
+      showComment: false,
+    },
+  });
+  return {
+    encryptedMessage: data.toString('base64'),
```
Comment (zhongxishen) [inline modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts:24]: See my comment about adding `AuthMessage`.

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  await verificationResult.signatures[0].verified;
+  return Buffer.from(decryptedMessage.data).toString('base64');
+}
+
+/**
+ * Verifies signature on a binary (message passed should be encoded in base64).
+ * @param signedMessage message to verify
+ * @param publicArmor public key to verify signature with
+ */
+export async function verifySignedData(signedMessage: AuthEncMessage, publicArmor: string): Promise<boolean> {
```
Comment (zhongxishen) [inline modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts:107]: See my comment about adding `AuthMessage`.

File: modules/sdk-lib-mpc/test/unit/tss/ecdsa/dklsDkg.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+      broadcastMessages: bitgoRound4Messages.broadcastMessages.concat(backupRound4Messages.broadcastMessages),
+    });
+    bitgo.handleIncomingMessages({
+      p2pMessages: [],
+      broadcastMessages: backupRound4Messages.broadcastMessages.concat(userRound4Messages.broadcastMessages),
+    });
+    backup.handleIncomingMessages({
+      p2pMessages: [],
+      broadcastMessages: bitgoRound4Messages.broadcastMessages.concat(userRound4Messages.broadcastMessages),
+    });
```
Comment (zhongxishen) [inline modules/sdk-lib-mpc/test/unit/tss/ecdsa/dklsDkg.ts:76]: maybe add a check that all three parties get to the same public key in their keyshares.

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/types.ts (hunk 2 above)
Comment (islamaminBitGo) [reply modules/sdk-lib-mpc/src/tss/ecdsa-dkls/types.ts]: done

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts (hunk 3 above)
Comment (islamaminBitGo) [reply modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts]: done

File: modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts (hunk 4 above)
Comment (islamaminBitGo) [reply modules/sdk-lib-mpc/src/tss/ecdsa-dkls/commsLayer.ts]: done

File: modules/sdk-lib-mpc/test/unit/tss/ecdsa/dklsDkg.ts (hunk 5 above)
Comment (islamaminBitGo) [reply modules/sdk-lib-mpc/test/unit/tss/ecdsa/dklsDkg.ts:77]: done

**Review by zhongxishen — APPROVED** [review]
LGTM

**Review by mmcshinsky-bitgo — APPROVED** [review]

**Pull request merged**